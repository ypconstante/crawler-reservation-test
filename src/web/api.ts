import fastify from 'fastify'
import {Server as HttpServer, IncomingMessage, ServerResponse} from 'http'

export type Server = fastify.FastifyInstance<HttpServer, IncomingMessage, ServerResponse>
