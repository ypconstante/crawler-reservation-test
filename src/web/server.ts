import fastify from 'fastify'
import {Server} from './api'
import {closeLogger, log} from '../core/log'
import 'source-map-support/register'

let server: Server = null

export async function startServer() {
  server = createServer()

  try {
    const port = 3000
    await server.listen(port)
    return server
  } catch (err) {
    log.error(err)
    process.exit(1)
    throw err
  }
}

export function createServer() {
  const server = fastify({
    logger: log,
  })

  require('./rest/search').register(server)

  return server
}

export function stopServer() {
  closeLogger()
  return new Promise(resolve => {
    server.close(
      () => resolve(),
    )
  })
}

