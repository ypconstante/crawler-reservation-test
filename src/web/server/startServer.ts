import fastify from 'fastify'
import {Server} from '../api'
import {registerSearch} from '../rest/search'
import {log} from '../../core/log'

export async function startServer() {
  const server: Server = fastify({
    logger: log,
  })

  registerSearch(server)

  try {
    const port = 3000
    await server.listen(port)
    return server
  } catch (err) {
    log.error(err)
    process.exit(1)
    throw err
  }
}
