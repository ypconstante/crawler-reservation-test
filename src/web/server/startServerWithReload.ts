import chokidar from 'chokidar'
import {log} from '../../core/log'
import {Server} from '../api'

let timeoutId: NodeJS.Timer = null
let serverPromise: Promise<Server> = null

export function startServerWithReload() {
  const watcher = chokidar.watch('./out/dist')

  watcher.on('ready', () => {
    log.info('Starting server with reload')

    serverPromise = startServer()

    watcher.on('all', scheduleResetServer)
  })
}

function startServer(): Promise<Server> {
  return require('./startServer').startServer()
}


function scheduleResetServer() {
  clearTimeout(timeoutId)
  timeoutId = setTimeout(resetServer, 50)

  function resetServer(): void {
    log.info('Reseting server')
    serverPromise.then(
      server => server.close(
        () => {
          clearAppCache()
          serverPromise = startServer()
        },
      ),
    )
  }

  function clearAppCache() {
    Object.keys(require.cache)
      .filter(shouldBeClearedFromCache)
      .forEach((path) => delete require.cache[path])
  }

  function shouldBeClearedFromCache(path: string) {
    return !isLogFile(path) && isAppFile(path)
  }

  function isLogFile(path: string) {
    return /[\/\\]core[\/\\]log.js$/.test(path)
  }

  function isAppFile(path: string) {
    return /[\/\\]out[\/\\]dist[\/\\]/.test(path)
  }

}
