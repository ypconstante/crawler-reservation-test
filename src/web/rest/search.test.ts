import {noop} from 'lodash'
import {Server} from '../api'
import {createServer} from '../server'

let server: Server

beforeEach(() => {
  server = createServer()
})

afterEach(() => {
  // FIXME fastify typings are declaring callback as mandatory
  server.close(noop)
})

test('GET /buscar', async () => {
  const response = await server.inject({
    method: 'GET',
    url: '/buscar',
  })

  expect(response.statusCode).toBe(200)
  expect(response.payload).toMatchSnapshot()
})
