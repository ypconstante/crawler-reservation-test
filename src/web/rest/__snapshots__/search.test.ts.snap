// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /buscar 1`] = `"{\\"url\\":\\"https://www.youtube.com/watch?v=dQw4w9WgXcQ\\"}"`;
